web/packages/web-client/server.js

```diff
@@ -3,11 +3,48 @@ const { parse } = require('url');
 const next = require('next');
 const fs = require('fs');
 
-const app = next({ dev: process.env.NODE_ENV !== 'production' });
+const isDev = process.env.NODE_ENV !== 'production';
+
+const app = next({ dev: isDev });
 const handle = app.getRequestHandler();
 
 const favicon = fs.readFileSync('./static/favicon.ico');
 
+const cache = {};
+
+const send = (res, html) => {
+  res.writeHeader(200, 'text/html');
+  res.write(html);
+  res.end();
+};
+
+const pagesToCache = {
+  '/': isDev ? 0 : -1, // forever
+  '/scorecard': isDev ? 0 : 3 * 60 * 60 * 1000 // 3 min,
+};
+
+const render = (cacheTime = 5 * 60 * 1000, req, res, pagePath, query) => {
+  if (
+    !!cache[req.url] &&
+    (cache[req.url].expires === -1 || cache[req.url].expires >= Date.now())
+  ) {
+    send(res, cache[req.url].html);
+    return;
+  }
+
+  app
+    .renderToHTML(req, res, pagePath, query)
```

Comment:
So far it appears that the way I'm passing authentication state between pages won't result in leaked state as a result of caching, but worth keeping an eye on.

This results in the fastest performance possible without significant deviations from the stack. For instance, in the case of scorecard, the Next/React SSR performance appears to be something like 10ms overhead above the cost of the request fetch, which is quite comparable to our old Flask solution, if not somewhat better.

Using this basic caching strategy, load drops to something like 1.5ms + your connection latency (including of course TLS negotiation), which is much better than either solution, and that of course makes sense, since, it's just static html, no interpolation whatsoever.